src/Sylius/Bundle/ApiBundle/Controller/TokenController.php

```diff
@@ -0,0 +1,78 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Bundle\ApiBundle\Controller;
+
+use FOS\RestBundle\Decoder\DecoderProviderInterface;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\ParameterBag;
+use OAuth2\OAuth2;
+use OAuth2\OAuth2ServerException;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpKernel\Exception\BadRequestHttpException;
+
+/**
+ * @author Łukasz Chruściel <lchrusciel@gmail.com>
+ */
+final class TokenController
+{
+    /**
+     * @var OAuth2
+     */
+    protected $server;
+
+    /**
+     * @var DecoderProviderInterface
+     */
+    private $decoderProvider;
+
+    /**
+     * @param OAuth2 $server
+     * @param DecoderProviderInterface $decoderProvider
+     */
+    public function __construct(OAuth2 $server, DecoderProviderInterface $decoderProvider)
+    {
+        $this->server = $server;
+        $this->decoderProvider = $decoderProvider;
+    }
+
+    /**
+     * @param Request $request
+     *
+     * @return Response
+     */
+    public function tokenAction(Request $request)
+    {
+        $contentType = $request->headers->get('Content-Type');
+
+        $format = is_null($contentType) ? $request->getRequestFormat() : $request->getFormat($contentType);
+
+        $content = $request->getContent();
+
+        if ($this->decoderProvider->supports($format)) {
+            if (!empty($content)) {
+                $decoder = $this->decoderProvider->getDecoder($format);
+                $data = $decoder->decode($content);
+                if (is_array($data)) {
+                    $request->request = new ParameterBag($data);
+                } else {
+                    throw new BadRequestHttpException('Invalid '.$format.' message received');
+                }
+            }
+        }
```

Comment:
Can we extract previous code to a private method? It will make this code intention clearer, especially with `@see` annotation mentioned previously.